cmd/codegen/generator/go/templates/modules.go

```diff
@@ -319,49 +356,51 @@ func (ps *parseState) fillObjectFunctionCases(type_ types.Type, cases map[string
 
 		fnCallArgs := []Code{Op("&").Id(parentVarName)}
 
-		for i := 0; i < sig.Params().Len(); i++ {
-			arg := sig.Params().At(i)
-
-			if i == 0 && arg.Type().String() == "context.Context" {
+		vars := map[string]struct{}{}
+		for i, spec := range method.paramSpecs {
+			if i == 0 && spec.paramType.String() == contextTypename {
 				fnCallArgs = append(fnCallArgs, Id("ctx"))
 				continue
 			}
 
-			if opts, ok := namedOrDirectStruct(arg.Type()); ok {
-				optsName := arg.Name()
-
-				statements = append(statements,
-					Var().Id(optsName).Id(renderNameOrStruct(arg.Type())))
-
-				for f := 0; f < opts.NumFields(); f++ {
-					param := opts.Field(f)
-
-					argName := strcase.ToLowerCamel(param.Name())
-					statements = append(statements,
-						If(Id(inputArgsVar).Index(Lit(argName)).Op("!=").Nil()).Block(
-							Err().Op("=").Qual("json", "Unmarshal").Call(
-								Index().Byte().Parens(Id(inputArgsVar).Index(Lit(argName))),
-								Op("&").Id(optsName).Dot(param.Name()),
-							),
-							checkErrStatement,
-						))
+			var target *Statement
+			if spec.parent == nil {
+				argName := strcase.ToLowerCamel(spec.name)
+				if _, ok := vars[argName]; !ok {
+					tp, access := findOptsAccessPattern(spec.paramType, Id(argName))
+					statements = append(statements, Var().Id(argName).Id(renderNameOrStruct(tp)))
+					if spec.variadic {
+						fnCallArgs = append(fnCallArgs, access.Op("..."))
+					} else {
+						fnCallArgs = append(fnCallArgs, access)
+					}
+					vars[argName] = struct{}{}
 				}
 
-				fnCallArgs = append(fnCallArgs, Id(optsName))
+				target = Id(argName)
 			} else {
-				argName := strcase.ToLowerCamel(arg.Name())
+				if _, ok := vars[spec.parent.name]; !ok {
+					tp, access := findOptsAccessPattern(spec.parent.paramType, Id(spec.parent.name))
+					statements = append(statements, Var().Id(spec.parent.name).Id(renderNameOrStruct(tp)))
+					if spec.variadic {
```

Comment:
nit, mostly for my own understanding: the inline struct can't be variadic, right? Just wondering if this case would ever get hit.

Reply:
Nope, this case isn't possible. However, the repetition in this part of the code *is* unclear, so I've refactored to have just one call to `findOptsAccessPattern` and only one check of `spec.variadic`.